modules/030-cloud-provider-yandex/hooks/yandex_preemptibly_delete_preemtible_instances.go

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright 2022 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	preemtibleVMDeletionDuration = 24 * time.Hour
+)
+
+type Machine struct {
+	Name              string
+	CreationTimestamp metav1.Time
+	Terminating       bool
+	MachineClassKind  string
+	MachineClassName  string
+}
+
+type YandexInstanceClass struct {
+	Name        string
+	Preemptible bool
+}
+
+func applyMachineFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var terminating bool
+	if obj.GetDeletionTimestamp() != nil {
+		terminating = true
+	}
+
+	classKind, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "class", "kind")
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(classKind) == 0 {
+		return nil, fmt.Errorf("spec.class.kind is empty in %q", obj.GetName())
+	}
+
+	className, _, err := unstructured.NestedString(obj.UnstructuredContent(), "spec", "class", "name")
+	if err != nil {
+		return nil, fmt.Errorf("can't access class name of Machine %q: %s", obj.GetName(), err)
+	}
+	if len(className) == 0 {
+		return nil, fmt.Errorf("spec.class.name is empty in %q", obj.GetName())
+	}
+
+	return &Machine{
+		Name:              obj.GetName(),
+		CreationTimestamp: obj.GetCreationTimestamp(),
+		Terminating:       terminating,
+		MachineClassKind:  classKind,
+		MachineClassName:  className,
+	}, nil
+}
+
+func isPreemptibleFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	preemptible, ok, err := unstructured.NestedBool(obj.UnstructuredContent(), "spec", "preemptible")
+	if err != nil {
+		return nil, fmt.Errorf("can't access field \"preemptible\" of YandexInstanceClass %q: %s", obj.GetName(), err)
+	}
+
+	preemptible = ok && preemptible
```

Comment:
In this case, it is possible to return `nil, nil` from the filter function and filter nils later (we use this pattern in other hooks). The same goes for the machines snapshot.